# CIFAR10 EfficientNetB7-ResidualMLP Neural Architecture Search

## Evaluation of best model from 2022-02-12_14_48 run

### For a description of the ResidualMLP and my neural architecture search algorithym, refer to: https://github.com/david-thrower/residual_MLP_Tests

## The following cells are the NAS and hyperparameter search task run:

File 1: ### The task trigger, 'ultraparameters', and hyperparameter search space for the task:

In [3]:
! cat task_trigger.py

"""Runs task.py, is a more convenient way to set up the parameters than
manually writing a shell script."""

import subprocess
preliminary_install_commands = ["pip3 install --upgrade pip",
                                "pip3 install pendulum"]
for cmd in preliminary_install_commands:
    subprocess.run(cmd,
                   shell=True,
                   check=True)
import pendulum

# Configure the run with this dict.
# Always enter boolean and floats as strings.

TIME_STAMP = pendulum.now().\
                    	__str__().\
                    	replace('T','_').\
                    	replace(':','-')[:16]

HPARAMS = {
    'project_name':"CIFAR10_EfficientNetB7-ResidualMLP_NAS_THIRD_PASS_SEARCH",
    'problem_type':"classification",
    'number_of_classes':10,
    'training_set_size':50000,
    'patience':25,
    'patience_min_delta':0.00001,
    'batch_size':50,
    'max_epochs':150,
    'minimum_learning_rate':0.0001,
    'maximum_learning_rate':0.02,
    'number_of_learning_rat

## File 2: The parameterized training task:

In [4]:
! cat task.py

import os
import subprocess
import argparse
# Add parser for common params

try:
    import pendulum
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pendulum", 
    	           shell=True, 
    	           check=True)
    import pendulum

try:
    import pandas as pd
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pandas", 
    	           shell=True, 
    	           check=True)
    import pandas as pd

try:
    import keras_tuner as kt
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install keras_tuner", 
    	           shell=True, 
    	           check=True)
    import keras_tuner as kt


try:
    import tensorflow as

## File 3: My python package > module residualmlp.residual_mlp that was used to run the neural architcture search task.

In [6]:
! cat residualmlp/residual_mlp.py

try:
    import keras_tuner as kt
except Exception as exc:
    print("Importing Keras tuner appears to be unsuccesful. "
          "keras tuner may need to be installed $ pip install -q -U "
          "keras-tuner. The auto-ml features are disabled until this is "
           "fixed, but ResidualMLP will work. A more detailed error is: "
           f"{exc}")
import tensorflow as tf
import numpy as np
import pandas as pd
import pendulum

# Becomes a layer to convert BW images to RGB

class ResidualMLP:
    
    def __init__(self, problem_type = 'classification',
                      learning_rate = .0007,
                      minimum_learning_rate = 0.00007,
                      maximum_learning_rate = 0.7,
                      number_of_learning_rates_to_try = 5,
                      input_shape = (32,32,3),
                      bw_images = False,
                      base_model = '',
                      base_model_input_shape = (600,600,3),
                      base_model_hyp

## File 4: The first 150 lines from the console logs from the task run: 

In [13]:
!cat 2022-02-12_14-48_CIFAR10_EfficientNetB7-ResidualMLP_NAS_SECOND_PASS_SEARCH_python3_shell_log.txt | head -n 200

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.7/11.7 MB 21.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.0/98.0 KB 14.4 MB/s eta 0:00:00
We're missing a module:
No module named 'pandas'
No problem, we install it...
We're missing a module:
No module named 'keras_tuner'
No problem, we install it...
# To reproduce the environment the NAS task was run in, copy and paste the following information into the evaluation script environent (excluding this line). PLEASE NOTE!!: If you do not reporduce these variables, the hyperband will not reproduce the correct model!
Date = 2022-02-12_14_48
PROBLEM_TYPE = classification
NUMBER_OF_CLASSES = 10
INPUT_SHAPE = (32, 32, 3)
BASE_MODEL_INPUT_SHAPE = (600, 600, 3)
PROJECT_NAME = CIFAR10_EfficientNetB7-ResidualMLP_NAS_THIRD_PASS_SEARCH
TRAINING_SET_SIZE = 50000
PATIENCE = 25
PATIENCE_MIN_DELTA = 1e-05
BATCH_SIZE = 50
MAX_EPOCHS = 150
RESULTS_DIR_FOR_SEARCH = 2022-02-12_14_48_CIFAR10_EfficientNetB7-ResidualMLP_NAS_THIRD_PASS_SEARC

## Now to evauluate the best model found by the task:

### First make sure the container we are evaluating this in has all its dependencies:

In [7]:
import os
import subprocess
import argparse
# Add parser for common params

try:
    import pendulum
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pendulum", 
    	           shell=True, 
    	           check=True)
    import pendulum

try:
    import pandas as pd
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pandas", 
    	           shell=True, 
    	           check=True)
    import pandas as pd

try:
    import keras_tuner as kt
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install keras_tuner", 
    	           shell=True, 
    	           check=True)
    import keras_tuner as kt


try:
    import tensorflow as tf
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install tensorflow", 
    	           shell=True, 
    	           check=True)
    import tensorflow as tf

from residualmlp.residual_mlp import ResidualMLP

We're missing a module:
No module named 'pendulum'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


We're missing a module:
No module named 'pandas'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


We're missing a module:
No module named 'keras_tuner'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


## Second: Reproduce the environment the task was run in: (This is a crirical step. Some of the, what I like to call 'ultraparmeters' - hyperparameters that fundamentally alter the behavior of the HYPERmodel, not just the models it returns) must be set to the same settings as they were during training, otherwise, our call to Tuner.get_best_model() will NOT reproduce the correct model!

In [1]:
Date = '2022-02-12_14_48'
PROBLEM_TYPE = 'classification'
NUMBER_OF_CLASSES = 10
INPUT_SHAPE = (32, 32, 3)
BASE_MODEL_INPUT_SHAPE = (600, 600, 3)
PROJECT_NAME = 'CIFAR10_EfficientNetB7-ResidualMLP_NAS_THIRD_PASS_SEARCH'
TRAINING_SET_SIZE = 50000
PATIENCE = 25
PATIENCE_MIN_DELTA = 1e-05
BATCH_SIZE = 50
MAX_EPOCHS = 150
RESULTS_DIR_FOR_SEARCH = '2022-02-12_14_48_CIFAR10_EfficientNetB7-ResidualMLP_NAS_THIRD_PASS_SEARCH_SEARCH_RUN'
MINIMUM_LEARNING_RATE = 0.0001
MAXIMUM_LEARNING_RATE = 0.02
NUMBER_OF_LEARNING_RATES_TO_TRY = 7
MINIMUM_NUMBER_OF_BLOCKS = 2
MAXIMUM_NUMBER_OF_BLOCKS = 3
MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK = 2
MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK = 4
MINIMUM_NEURONS_PER_BLOCK_LAYER = 70
MAXIMUM_NEURONS_PER_BLOCK_LAYER = 140
N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY = 7
MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY = 15
MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY = 35
MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS = 0.2
MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS = 0.6
N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS = 3
MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK = 0
MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK = 85
N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK = 7
MINIMUM_DROPOUT_RATE = 0.2
MAXIMUM_DROPOUT_RATE = 0.6
N_OPTIONS_DROPOUT_RATE = 5
MINIMUM_FINAL_DENSE_LAYERS = 33
MAXIMUM_FINAL_DENSE_LAYERS = 300
N_OPTIONS_FINAL_DENSE_LAYERS = 7

## Third, reproduce the dataset:

In [16]:

cifar = tf.keras.datasets.cifar10.load_data()
(x_train, y_train), (x_test, y_test) = cifar

y_train_ohe = tf.one_hot([i[0] for i in  y_train],10)
indexes_for_rows = tf.range(0,y_train.shape[0])
shuffled_indexes = tf.random.shuffle(indexes_for_rows)
selected_indexes = shuffled_indexes[:TRAINING_SET_SIZE]
selected_x_train = x_train[selected_indexes,:,:,:]
selected_y_train_ohe = y_train_ohe.numpy()[selected_indexes,:]


170508288/170498071 [==============================] - 8s 0us/step


2022-02-13 21:10:08.938581: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-13 21:10:08.976250: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-13 21:10:08.977179: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-13 21:10:08.979524: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-13 21:10:08.980386: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read f

## Fourth, reproduce the EfficientNetB7 base model used for transfer learning

In [17]:
mod_with_fc_raw = tf.keras.applications.efficientnet.EfficientNetB7(
    include_top=True, weights='imagenet', input_tensor=None,
    input_shape = BASE_MODEL_INPUT_SHAPE, pooling='max', classes=1000
)

# Make the deepest conv2d layer trainable, leave everything else
# as not trainable
for layer in mod_with_fc_raw.layers:
    layer.trainable = False
# Last conv2d layer. This we want to train .
mod_with_fc_raw.layers[-6].trainable = True

# Create the final base model
# (remove the final Dense and BatchNormalization layers ...) 
efficient_net_b_7_transferable_base_model =    tf.keras.Model(inputs=mod_with_fc_raw.layers[0].input, 
                    outputs=mod_with_fc_raw.layers[-3].output)

268337152/268326632 [==============================] - 12s 0us/step


## Fifth: Reporduce the hypermodel and tuner used to perform the neural architecture and hyperparmaeter search:

In [18]:
model_builder = ResidualMLP(
                    problem_type = PROBLEM_TYPE,
                    minimum_learning_rate = MINIMUM_LEARNING_RATE, 
                    maximum_learning_rate = MAXIMUM_LEARNING_RATE, 
                    number_of_learning_rates_to_try =
                        NUMBER_OF_LEARNING_RATES_TO_TRY, 
                    input_shape = INPUT_SHAPE, 
                    bw_images = False, 
                    base_model = 
                        efficient_net_b_7_transferable_base_model, 
                    base_model_input_shape = BASE_MODEL_INPUT_SHAPE, 
                    flatten_after_base_model = False, 
                    minimum_number_of_blocks = MINIMUM_NUMBER_OF_BLOCKS, 
                    maximum_number_of_blocks = MAXIMUM_NUMBER_OF_BLOCKS, 
                    minimum_number_of_layers_per_block =
                        MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK, 
                    maximum_number_of_layers_per_block =
                        MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK,
                    minimum_neurons_per_block_layer =
                        MINIMUM_NEURONS_PER_BLOCK_LAYER, 
                    maximum_neurons_per_block_layer =
                        MAXIMUM_NEURONS_PER_BLOCK_LAYER, 
                    n_options_of_neurons_per_layer_to_try =
                        N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY, 
                    minimum_neurons_per_block_layer_decay =
                        MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY, 
                    maximum_neurons_per_block_layer_decay = 
                        MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY, 
                    minimum_dropout_rate_for_bypass_layers =
                        MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS, 
                    maximim_dropout_rate_for_bypass_layers =
                        MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS, 
                    n_options_dropout_rate_for_bypass_layers =
                        N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS,
                    minimum_inter_block_layers_per_block =
                        MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK, 
                    maximum_inter_block_layers_per_block =
                        MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK,
                    n_options_inter_block_layers_per_block =                        N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK,
                    minimum_dropout_rate = MINIMUM_DROPOUT_RATE, 
                    maximum_dropout_rate = MAXIMUM_DROPOUT_RATE,
                    n_options_dropout_rate = N_OPTIONS_DROPOUT_RATE, 
                    minimum_final_dense_layers =
                        MINIMUM_FINAL_DENSE_LAYERS,
                    maximum_final_dense_layers =
                        MAXIMUM_FINAL_DENSE_LAYERS, 
                    n_options_final_dense_layers =
                        N_OPTIONS_FINAL_DENSE_LAYERS, 
                    number_of_classes = NUMBER_OF_CLASSES,
                    final_activation = tf.keras.activations.softmax)



logdir_for_search = os.path.join("logs", RESULTS_DIR_FOR_SEARCH + "_TB")
tensorboard_callback_search =    tf.keras.callbacks.TensorBoard(logdir_for_search, histogram_freq=1)

tuner = kt.Hyperband(
    model_builder.build_auto_residual_mlp,
    objective='val_loss',
    project_name = PROJECT_NAME,
    max_epochs = MAX_EPOCHS,
    hyperband_iterations = 2)

2022-02-13 21:10:35.429367: I tensorflow/core/profiler/lib/profiler_session.cc:131] Profiler session initializing.
2022-02-13 21:10:35.429426: I tensorflow/core/profiler/lib/profiler_session.cc:146] Profiler session started.
2022-02-13 21:10:35.430607: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1614] Profiler found 1 GPUs


INFO:tensorflow:Reloading Oracle from existing project ./CIFAR10_EfficientNetB7-ResidualMLP_NAS_THIRD_PASS_SEARCH/oracle.json


2022-02-13 21:10:35.605893: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session tear down.
2022-02-13 21:10:35.606135: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1749] CUPTI activity buffer flushed


All permutations:
     number_of_blocks  layers_per_block  neurons_per_block_layer  \
0                   2                 2                       70   
1                   2                 2                       70   
2                   2                 2                       70   
3                   2                 2                       70   
4                   2                 2                       70   
..                ...               ...                      ...   
877                 3                 4                      140   
878                 3                 4                      140   
879                 3                 4                      140   
880                 3                 4                      140   
881                 3                 4                      140   

     neurons_per_block_layer_decay  
0                               15  
1                               16  
2                               17  
3                

## Get info about the best models from the search:

In [19]:
tuner.results_summary(10)

Results summary
Results in ./CIFAR10_EfficientNetB7-ResidualMLP_NAS_THIRD_PASS_SEARCH
Showing 10 best trials
Objective(name='val_loss', direction='min')
Trial summary
Hyperparameters:
learning_rate: 0.0001245791948180787
blocks: 79
bypass_layers_units: 81
inter_block_layers: 85
final_dense_layers: 166
b_norm_or_dropout_residual_bypass_layers: bnorm
dropout_rate_for_bypass_layers: 0.2
b_norm_or_dropout_last_layers: bnorm
dropout_rate: 0.2
tuner/epochs: 2
tuner/initial_epoch: 0
tuner/bracket: 4
tuner/round: 0
Score: 0.23214344680309296
Trial summary
Hyperparameters:
learning_rate: 0.0006127124845172436
blocks: 270
bypass_layers_units: 105
inter_block_layers: 0
final_dense_layers: 33
b_norm_or_dropout_residual_bypass_layers: bnorm
dropout_rate_for_bypass_layers: 0.4
b_norm_or_dropout_last_layers: bnorm
dropout_rate: 0.4
tuner/epochs: 2
tuner/initial_epoch: 0
tuner/bracket: 4
tuner/round: 0
Score: 0.2366037219762802
Trial summary
Hyperparameters:
learning_rate: 0.00014524140478609202
block

## Let's examine the best model:

In [29]:
best_model = tuner.get_best_models(1)[0]

All permutations:
     number_of_blocks  layers_per_block  neurons_per_block_layer  \
0                   2                 2                       70   
1                   2                 2                       70   
2                   2                 2                       70   
3                   2                 2                       70   
4                   2                 2                       70   
..                ...               ...                      ...   
877                 3                 4                      140   
878                 3                 4                      140   
879                 3                 4                      140   
880                 3                 4                      140   
881                 3                 4                      140   

     neurons_per_block_layer_decay  
0                               15  
1                               16  
2                               17  
3                

## It appears that the best model is:

Blocks option: [[2,105,32],[2,105,32]]. This means:
1. There are 2 ResidualMLP blocks of 2 block body layers.
2. Each block having a head layer Dense(105) within the body of the residual block and Dense(73) on the second layer.
3. The residual bypass path for both ResidualMLP blocks has one Dense(81) layer.
4. There is one Dense(85) layer in between the 2 residual blocks.
5. There is one Dense(166) layer after the last block before the output layer.
6. All sequential Dense layers were striated with BatchNormalization() layers. 
7. The learning rate was 0.00012457
8. It appears the model only needed to train for 2 epochs to reach the accuracy it did!


In [21]:
best_model.summary()

Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            [(None, 32, 32, 3)]  0                                            
__________________________________________________________________________________________________
resizing (Resizing)             (None, 600, 600, 3)  0           input_1[0][0]                    
__________________________________________________________________________________________________
model (Functional)              (None, 2560)         64097687    resizing[0][0]                   
__________________________________________________________________________________________________
dense (Dense)                   (None, 105)          268905      model[1][0]                      
______________________________________________________________________________________________

## Let's evalate the model:

In [30]:
y_test_ohe = tf.one_hot([i[0] for i in  y_test],10)
best_model.evaluate(x_test,y_test_ohe)

2022-02-13 21:33:50.882766: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)
2022-02-13 21:33:58.064038: I tensorflow/stream_executor/cuda/cuda_dnn.cc:381] Loaded cuDNN version 8204


313/313 [==============================] - 955s 3s/step - loss: 0.2548 - top_1_categorical_accuracy: 0.9280 - top_2_categorical_accuracy: 0.9750 - top_3_categorical_accuracy: 0.9882 - top_4_categorical_accuracy: 0.9934 - top_5_categorical_accuracy: 0.9960 - top_6_categorical_accuracy: 0.9973 - top_7_categorical_accuracy: 0.9983 - top_8_categorical_accuracy: 0.9990 - top_9_categorical_accuracy: 0.9995 - precision: 0.9365 - recall: 0.9220 - accuracy: 0.0000e+00


[0.25477176904678345,
 0.9279999732971191,
 0.9750000238418579,
 0.9882000088691711,
 0.993399977684021,
 0.9959999918937683,
 0.9973000288009644,
 0.9983000159263611,
 0.9990000128746033,
 0.9994999766349792,
 0.9365159869194031,
 0.921999990940094,
 0.0]

## It looks like I almost reached 93% top-1-categorical-accuracy: 0.9279999732971191

In [34]:
DATE = '2022-02-12_14_48' # Date of training task execution...
best_model.save(f'{DATE}_interim_best_model_EXPORTED_MODEL')

2022-02-13 22:25:13.475689: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.


INFO:tensorflow:Assets written to: 2022-02-12_14_48_interim_best_model_EXPORTED_MODEL/assets


/usr/local/lib/python3.8/dist-packages/keras/utils/generic_utils.py:494: CustomMaskWarning: Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.
  warnings.warn('Custom mask layers require a config and must override '
